src/main/java/xades4j/production/SignerBES.java

```diff
@@ -244,14 +248,27 @@ public final XadesSignatureResult sign(
             // with its value set to: http://uri.etsi.org/01903#SignedProperties."
 
             String digestAlgUri = algorithmsProvider.getDigestAlgorithmForDataObjsReferences();
-            if (null == digestAlgUri)
+            if (StringUtils.isNullOrEmptyString(digestAlgUri))
             {
                 throw new NullPointerException("Digest algorithm URI not provided");
             }
 
+            // Use same canonicalization URI as specified in the ds:CanonicalizationMethod for Signature.
+            String canonAlg = this.algorithmsProvider.getCanonicalizationAlgorithmForSignature().getUri();
+
             try
             {
-                signature.addDocument('#' + signedPropsId, null, digestAlgUri, null, QualifyingProperty.SIGNED_PROPS_TYPE_URI);
+                Transforms transforms = null;
+
+                if (!StringUtils.isNullOrEmptyString(canonAlg))
+                {
+                    // HACK: since we're not using Canonicalizer, do a quick check to ensure
+                    // that 'c14n' refers to a configured C14N algorithm.
+                    transforms = new Transforms(signatureDocument);
+                    transforms.addTransform(Canonicalizer.getInstance(canonAlg).getURI());

```

Comment:
I mean parameters of the algorithm itself, using AlgorithmParametersMarshaller. But I'll add this later. I've been picky enough :)